File: westcoast/src/main/java/frc/commands/BallHandlers/IndexLogic.java
package frc.commands.BallHandlers;

import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.subsystems.BallHandlingSubsystem;

public class IndexLogic extends CommandBase {
  private final BallHandlingSubsystem ballHandler;
  private boolean end;

  public IndexLogic(
    BallHandlingSubsystem ballHandler
  ) {
    this.ballHandler=ballHandler;
  }

  @Override
  public void initialize() {
  }

  @Override
  public void execute() {
    switch (ballHandler.getSensorSwitch())  {
      case 0x0: ballHandler.runsyncMagdex(0.5); //No sensors see anything.
        break;
      case 0x1: ballHandler.runsyncMagdex(0.5); //Intake sensor sees something.
        break;
      case 0x2: end = true; //Index sensor sees something.
        break;
      case 0x3: ballHandler.runsyncMagdex(0.5); //Both Index sensor and Intake sensor see something (all balls after first)
        break;
      case 0x4: end = true; //End sensor sees something (shouldn't happen).
        break;
      case 0x5: end = true; //Intake sensor and End sensor see something (shouldn't happen).
        break;
      case 0x6: end = true; //End sensor and Index sensor sees something (happens when 4 balls are in the system).
        break;
      case 0x7: end = true; //All sensors are saturated (happens with 5 balls).
        break;
    }
  }

  @Override
  public void end(boolean interrupted) {
  }

  @Override
  public boolean isFinished() {
    return end;
  }
}

File: westcoast/src/main/kotlin/frc/commands/magazine/IndexLogic.kt
package frc.commands.magazine

import edu.wpi.first.wpilibj2.command.CommandBase
import frc.subsystems.MagazineSubsystem

class IndexLogic(
    private val magazine: MagazineSubsystem
) : CommandBase() {

  private var boost = 0.0

  override fun initialize() {
    boost = 0.019 * magazine.indexCount
  }

  override fun execute() {
    when (magazine.sensorSwitch) {
      0x0 -> {
        magazine.runIndexMotor(0.3)
        magazine.runMagMotor(0.3 + boost) //No sensors see anything.
      }
      0x1 -> {
        magazine.runIndexMotor(0.3)
        magazine.runMagMotor(0.43) //Intake sensor sees something.
      }
      0x2 -> Unit //Index sensor sees something.
      0x3 -> {
        magazine.runIndexMotor(0.27) //Both Index sensor and Intake sensor see something (all balls after first)
        magazine.runMagMotor(0.3 + boost) //Boosts magazines speed so as to avoid magazine still seeing ball while index get's cleared (could also be fixed by moving sensor positions)
      }
      0x4 -> Unit //End sensor sees something (shouldn't happen).
      0x5 -> Unit //Intake sensor and End sensor see something (shouldn't happen).
      0x6 -> Unit //End sensor and Index sensor sees something (happens when 4 balls are in the system).
      0x7 -> Unit //All sensors are saturated (happens with 5 balls).
    }
  }

  override fun end(interrupted: Boolean) {
    println("Index Complete")
    magazine.stopMotors()
  }

  override fun isFinished(): Boolean =
      when (magazine.sensorSwitch) {
        0x0 -> false
        0x1 -> false
        0x2 -> true
        0x3 -> false
        0x4 -> true
        0x5 -> true
        0x6 -> true
        0x7 -> true
        else -> true
      }

  init {
    addRequirements(magazine)
  }
}